helix-term/src/ui/explorer.rs

```diff
@@ -0,0 +1,1400 @@
+            return on_next_key(cx, self, key_event);
+        }
+
+        if let EventResult::Consumed(c) = self.handle_prompt_event(key_event, cx) {
+            return EventResult::Consumed(c);
+        }
+
+        (|| -> Result<()> {
+            match key_event {
+                key!(Esc) => self.unfocus(),
+                key!('q') => self.close(),
+                key!('?') => self.toggle_help(),
+                key!('a') => self.new_create_file_or_folder_prompt(cx)?,
+                shift!('B') => self.change_root_parent_folder()?,
+                key!(']') => self.change_root_to_current_folder()?,
+                key!('[') => self.go_to_previous_root(),
+                key!('d') => self.new_remove_prompt()?,
+                key!('r') => self.new_rename_prompt(cx)?,
+                key!('-') | key!('_') => self.decrease_size(),
+                key!('+') | key!('=') => self.increase_size(),
+                _ => {
+                    self.tree
+                        .handle_event(&Event::Key(*key_event), cx, &mut self.state);
+                }
+            };
+            Ok(())
+        })()
+        .unwrap_or_else(|err| cx.editor.set_error(format!("{err}")));
+
+        EventResult::Consumed(None)
+    }
+
+    fn render(&mut self, area: Rect, surface: &mut Surface, cx: &mut Context) {
+        if area.width < 10 || area.height < 5 {
+            cx.editor.set_error("explorer render area is too small");
+            return;
+        }
+        let config = &cx.editor.config().explorer;
+        let position = config.position;
+        self.render_embed(area, surface, cx, &position);
+    }
+
+    fn cursor(&self, area: Rect, editor: &Editor) -> (Option<Position>, CursorKind) {
+        if let Some(prompt) = self
+            .prompt
+            .as_ref()
+            .map(|(_, prompt)| prompt)
+            .or_else(|| self.tree.prompt())
+        {
+            let (x, y) = (area.x, area.y + area.height.saturating_sub(1));
+            prompt.cursor(Rect::new(x, y, area.width, 1), editor)
+        } else {
+            (None, CursorKind::Hidden)
+        }
+    }
+}
+
+fn render_block(area: Rect, surface: &mut Surface, borders: Borders) -> Rect {
+    let block = Block::default().borders(borders);
+    let inner = block.inner(area);
+    block.render(area, surface);
+    inner
+}
+
+#[cfg(test)]
+mod test_explorer {
+    use crate::compositor::Component;
+
+    use super::Explorer;
+    use helix_view::graphics::Rect;
+    use std::{fs, path::PathBuf};
+
+    /// This code should create the following file tree:
+    ///
+    ///   test_explorer/<name>
+    ///   ├── index.html
+    ///   ├── .gitignore
+    ///   ├── scripts
+    ///   │   └── main.js
+    ///   └── styles
+    ///      ├── style.css
+    ///      └── public
+    ///          └── file
+    ///
+    fn dummy_file_tree(name: &str) -> PathBuf {
+        let path: PathBuf = format!("test_explorer{}{}", std::path::MAIN_SEPARATOR, name).into();
```

Comment:
This should not be making a directory in the same source tree; it should instead be making a temporary directory that only lives for the duration of the test. We use the [`tempfile`](https://docs.rs/tempfile/latest/tempfile/) crate for this in the integration tests.

Reply:
Resolved by https://github.com/helix-editor/helix/pull/5768/commits/2d1ca230809db637d743055e31f98e3d6889cb65